src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.JsonNetSyntaxChecks.cs

```diff
@@ -0,0 +1,236 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.Json
+{
+    using static EmbeddedSyntaxHelpers;
+
+    using JsonToken = EmbeddedSyntaxToken<JsonKind>;
+
+    internal partial struct JsonParser
+    {
+        private static class JsonNetSyntaxChecker
+        {
+            private static EmbeddedDiagnostic? CheckChildren(JsonNode node)
+            {
+                foreach (var child in node)
+                {
+                    if (child.IsNode)
+                    {
+                        var diagnostic = CheckSyntax(child.Node);
+                        if (diagnostic != null)
+                        {
+                            return diagnostic;
+                        }
+                    }
+                }
+
+                return null;
+            }
+
+            public static EmbeddedDiagnostic? CheckSyntax(JsonNode node)
+            {
+                switch (node.Kind)
+                {
+                    case JsonKind.Array: return CheckArray((JsonArrayNode)node);
+                    case JsonKind.Object: return CheckObject((JsonObjectNode)node);
+                    case JsonKind.Constructor: return CheckConstructor((JsonConstructorNode)node);
+                    case JsonKind.Property: return CheckProperty((JsonPropertyNode)node);
+                    case JsonKind.Literal: return CheckLiteral((JsonLiteralNode)node);
+                    case JsonKind.NegativeLiteral: return CheckNegativeLiteral((JsonNegativeLiteralNode)node);
+                }
+
+                return CheckChildren(node);
+            }
+
+            private static EmbeddedDiagnostic? CheckLiteral(JsonLiteralNode node)
+            {
+                if (node.LiteralToken.Kind == JsonKind.NumberToken)
+                {
+                    return CheckNumber(node.LiteralToken);
+                }
+
+                return CheckChildren(node);
+            }
+
+            private static EmbeddedDiagnostic? CheckNegativeLiteral(JsonNegativeLiteralNode node)
+            {
+                if (node.LiteralToken.Kind == JsonKind.NumberToken)
+                {
+                    return CheckNumber(node.LiteralToken);
+                }
+
+                return CheckChildren(node);
+            }
+
+            private static EmbeddedDiagnostic? CheckNumber(JsonToken numberToken)
+            {
+                // This code was effectively copied from:
+                // https://github.com/JamesNK/Newtonsoft.Json/blob/993215529562866719689206e27e413013d4439c/Src/Newtonsoft.Json/JsonTextReader.cs#L1926
+                // So as to match Newtonsoft.Json's behavior around number parsing.
```

Comment:
Given we're both MIT licensed I guess we can pull this code in; does this need some extra attribution/third party update though?

Reply:
good question.  @sharwell ?